src/common/primitive_exec_types.cpp

```diff
@@ -24,6 +24,44 @@
 namespace dnnl {
 namespace impl {
 
+memory_arg_t::memory_arg_t(memory_t *mem, bool is_const)
+    : mem_(mem), is_const_(is_const) {
+    if (mem_) mem_->retain();
+}
+
+memory_arg_t::~memory_arg_t() {
+    if (mem_) mem_->release();
+}
+
+memory_arg_t::memory_arg_t(const memory_arg_t &other)
+    : mem_(other.mem_), is_const_(other.is_const_) {
+    if (mem_) mem_->retain();
+}
+
+memory_arg_t &memory_arg_t::operator=(const memory_arg_t &other) {
+    if (this == &other) return *this;
+
+    mem_ = other.mem_;
+    is_const_ = other.is_const_;
+    if (mem_) mem_->retain();
+
+    return *this;
+}
+
+memory_arg_t::memory_arg_t(memory_arg_t &&other) {
+    *this = std::move(other);
+}
+
+memory_arg_t &memory_arg_t::operator=(memory_arg_t &&other) {
+    if (this == &other) return *this;
+
+    // Note: swapping memory objects lets dtor to handle original `mem_`.
+    std::swap(mem_, other.mem_);
+    is_const_ = other.is_const_;
```

Comment:
Maybe handle the logic explicitly to keep it simple? Something along the lines:

```cpp
    memory_deleter_t{}(mem_);

    mem_= other.mem_;
    is_const_= other.is_const_;

    other.mem_ = nullptr;

    return *this;
```

Reply:
Used the proposed version.